### PR TITLE
Decouples credentials logic from promotions

### PR DESCRIPTION
**Pull request by NejcZdovc**
Resolves https://github.com/brave/brave-browser/issues/8677

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [ ] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:
Plan 1:
- enable rewards
- claim promotion
- make sure that balance updates
- do a tip
- make sure that balance updates

Plan 2:
- make sure that VG claim to uphold works

Plan 3:
- enable rewards
- claim promotion
- close browser
- go into db and change `expires_at` of `promotion` table in the past, go to `unblinded_tokens` table and change `expires_at` to be in the past
- start browser
- make sure that balance is 0 and promotion expired

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: vendor/bat-native-ledger/src/bat/ledger/internal/credentials/credentials.h [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+      const CredentialsTrigger& trigger,
+      ledger::ResultCallback callback);
+
+  virtual void BlindedCreds(
+      const ledger::Result result,
+      const std::string& blinded_creds_json,
+      const CredentialsTrigger& trigger,
+      ledger::ResultCallback callback);
+
+  virtual void ClaimCreds(
```
Comment (tmancey) [inline vendor/bat-native-ledger/src/bat/ledger/internal/credentials/credentials.h]: nit: do we need `Creds` as the class is called `credentials`?

File: vendor/bat-native-ledger/src/bat/ledger/internal/credentials/credentials.h [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+class Credentials {
+ public:
+  virtual ~Credentials() = default;
+
+  virtual void Start(
+      const CredentialsTrigger& trigger,
+      ledger::ResultCallback callback);
+
+  virtual void BlindedCreds(
```
Comment (tmancey) [inline vendor/bat-native-ledger/src/bat/ledger/internal/credentials/credentials.h]: nit: do we need `Creds` as the class is called `credentials`?

File: vendor/bat-native-ledger/src/bat/ledger/internal/credentials/credentials_common.cc [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+using challenge_bypass_ristretto::BlindedToken;
+using challenge_bypass_ristretto::PublicKey;
+using challenge_bypass_ristretto::SignedToken;
+using challenge_bypass_ristretto::UnblindedToken;
+
+namespace {
+
+void ParseSignedCredsResponse(
+    const std::string& response,
+    base::Value* result) {
```
Comment (tmancey) [inline vendor/bat-native-ledger/src/bat/ledger/internal/credentials/credentials_common.cc]: `DCHECK` result

Comment (tmancey) [inline vendor/bat-native-ledger/src/bat/ledger/internal/credentials/credentials_common.cc]: `DCHECK` result

File: vendor/bat-native-ledger/src/bat/ledger/internal/credentials/credentials_common.h [hunk 4]
```diff
@@ -0,0 +1,57 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVELEDGER_CREDENTIALS_COMMON_H_
+#define BRAVELEDGER_CREDENTIALS_COMMON_H_
+
+#include <string>
```
Comment (tmancey) [inline vendor/bat-native-ledger/src/bat/ledger/internal/credentials/credentials_common.h]: `#include <stdint.h>` as using `uint64_t` etc.

File: vendor/bat-native-ledger/src/bat/ledger/internal/database/database.cc [hunk 5]
```diff
@@ -166,6 +167,33 @@ void Database::DeleteContributionQueue(
...
+    ledger::CredsBatchPtr info,
+    ledger::ResultCallback callback) {
+  creds_batch_->SaveSignedCreds(std::move(info), callback);
+}
+
+void Database::GetAllCredsBatches(ledger::GetAllCredsBatchCallback callback) {
+  creds_batch_->GetAllRecords(callback);
+}
+
+
```
Comment (tmancey) [inline vendor/bat-native-ledger/src/bat/ledger/internal/database/database.cc]: super nit: extra space

File: vendor/bat-native-ledger/src/bat/ledger/internal/credentials/credentials_common.cc (hunk 3 above)
Comment (NejcZdovc) [reply vendor/bat-native-ledger/src/bat/ledger/internal/credentials/credentials_common.cc]: fixed

File: vendor/bat-native-ledger/src/bat/ledger/internal/database/database.cc (hunk 5 above)
Comment (NejcZdovc) [reply vendor/bat-native-ledger/src/bat/ledger/internal/database/database.cc]: fixed

File: vendor/bat-native-ledger/src/bat/ledger/internal/credentials/credentials.h (hunk 1 above)
Comment (NejcZdovc) [reply vendor/bat-native-ledger/src/bat/ledger/internal/credentials/credentials.h]: done

File: vendor/bat-native-ledger/src/bat/ledger/internal/credentials/credentials.h (hunk 2 above)
Comment (NejcZdovc) [reply vendor/bat-native-ledger/src/bat/ledger/internal/credentials/credentials.h]: done

File: vendor/bat-native-ledger/src/bat/ledger/internal/credentials/credentials_common.h (hunk 4 above)
Comment (NejcZdovc) [reply vendor/bat-native-ledger/src/bat/ledger/internal/credentials/credentials_common.h]: done

**Comment by NejcZdovc** [comment]
CI passed on  android, linux and window

**Comment by NejcZdovc** [comment]
macOS and iOS failing because of https://github.com/brave/brave-browser/issues/8756

**Pull request merged**